### PR TITLE
Fix spring-integration sample

### PR DESCRIPTION
**Pull request by olegz**
- Added additional proxy checks
- Added discoverFunctionTypeFromClass() method to FunctionTypeUtils

### CODE REVIEW

File: spring-cloud-function-samples/function-sample-spring-integration/src/main/java/example/FunctionSampleSpringIntegrationApplication.java [hunk 1]
```diff
@@ -26,21 +26,21 @@
...
-	//@Bean
+	@Bean
 	public IntegrationFlow uppercaseFlow() {
 		return IntegrationFlows.from(MessageFunction.class, "uppercase")
 				.<String, String>transform(String::toUpperCase)
 				.logAndReply(LoggingHandler.Level.WARN);
 	}
 
-	public interface MessageFunction extends Function<Message<String>, Message<String>> {
+	public interface MessageFunction<T> extends Function<Message<String>, Message<String>> {
```
Comment (artembilan) [inline spring-cloud-function-samples/function-sample-spring-integration/src/main/java/example/FunctionSampleSpringIntegrationApplication.java]: Why do we need this `<T>` here?

Comment (olegz) [reply spring-cloud-function-samples/function-sample-spring-integration/src/main/java/example/FunctionSampleSpringIntegrationApplication.java]: No reason, just wanted to create a deep hierarchy and validate that the new `discover` method still finds the function type.

Comment (artembilan) [reply spring-cloud-function-samples/function-sample-spring-integration/src/main/java/example/FunctionSampleSpringIntegrationApplication.java]: OK, but it must not be as a part of a sample. It is going to confuse end-users.

Comment (olegz) [reply spring-cloud-function-samples/function-sample-spring-integration/src/main/java/example/FunctionSampleSpringIntegrationApplication.java]: ohh, i forgot to delete it. Good point

**Comment by artembilan** [comment]
@olegz ,

I want to share with you another point of view for this problem.

Instead of assuming about proxies, how about to perform a `registry.getBean("&" + actualName, FactoryBean.class).getObjectType()` logic in the `FunctionContextUtils.findType()`, when `FactoryBean.class.isAssinableFrom(definition.getBeanClass())` and call the same new `FunctionTypeUtils.discoverFunctionTypeFromClass()`?

**Pull request merged**